### PR TITLE
gracefully handle apps without `package.json` files (fixes #50)

### LINKED ISSUE #50: make it optional to have package.json file
#40 makes it optional to specify a main entry script in your package.json file (by defaulting index.js in that case). We could go further and make it optional to provide a package.json file at all, since none of its properties are necessary when you're running an app. (Packaging is somewhat different, since you presumably want your app to have a name, although even there we could fall back to a default name if one isn't provided.)

### PR DESCRIPTION
**Pull request by cvan**
@mykmelez: could use some help on the failing tests. (I typically use different test runners, so the whole `SIGINT`, `stdout`/`stderror`, streams, Promise chain is a bit unfamiliar to me.)

### CODE REVIEW

**Comment by mykmelez** [comment]
> @mykmelez: could use some help on the failing tests. (I typically use different test runners, so the whole SIGINT, stdout/stderror, streams, Promise chain is a bit unfamiliar to me.)

I've submitted https://github.com/cvan/qbrt/pull/1 to fix the test failures. I'm not entirely sure about some of the fixes—I may want to revisit them and do something else instead—but that PR will at least unblock this PR.

**Comment by cvan** [comment]
many thanks, much appreciated. will address feedback and set up Travis CI on my fork. I agree re: the warnings.

**Comment by cvan** [comment]
@mykmelez: for some reason [Taskcluster](https://taskcluster.github.io/unified-logviewer/?url=https%3A%2F%2Fqueue.taskcluster.net%2Fv1%2Ftask%2FXd4JYquTTx6U_eooYPkoZQ%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log&highlightStart=&highlightEnd=&lineNumber=&wrapLines=false&showLineNumbers=true&jumpToHighlight=true&followLog=false&lineHeight=19&asText=false) is failing on `test/run-path-missing-package.js` on Linux, but it seems to be fine on Windows and on Travis CI

**Comment by mykmelez** [comment]
> @mykmelez: for some reason Taskcluster is failing on test/run-path-missing-package.js on Linux, but it seems to be fine on Windows and on Travis CI (and locally)

This error message in the log identifies the failure:

> (node:1885) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 2): TypeError: Path must be a string. Received { qbrt: './bin/cli.js' }

When an app is missing a package.json manifest, qbrt now traverses its directory tree looking for one. And if that app is in a subdirectory of the qbrt directory (f.e. test/hello-world-missing-package/), then it find's qbrt's package.json manifest.

And that manifest specifies `"bin": "./bin/cli.js"`, which [normalize-package-data](https://www.npmjs.com/package/normalize-package-data) converts into something like `"bin": { qbrt: './bin/cli.js' }`, per this rule:

> If bin field is a string, then bin field will become an object with name set to the value of the name field, and bin set to the original string value.

Meanwhile, cli.js sets *appPackageResult.pkg.main* to `appPackageResult.pkg.main || appPackageResult.pkg.bin || 'index.js'`; and since qbrt's manifest doesn't have a *main* property, cli.js sets it to the *bin* property, which is the aforementioned object, triggering the error.

I tried to work around this issue in https://github.com/mozilla/qbrt/pull/83/commits/283bbc934010248854333f4496e892789a2b1698 by copying test/hello-world-missing-package/ to the temporary directory and then running it from there. But the temporary directory appears to be `/repo` in the TaskCluster Docker image, which is presumably also the directory where it clones qbrt, thus foiling my ingenious workaround. ;-)

So I'm not sure what to do next. One option would be to find some other directory where we can copy the app on Linux (at least in TaskCluster, but ideally we'd have a general solution). Alternately, we could check *appPackageResult.pkg.bin* to ensure it's a reasonable value before trying to use it. Or we could remove the read-pkg-up functionality, although presumably it's useful, so that's a last resort.

**Comment by mykmelez** [comment]
> So I'm not sure what to do next. One option would be to find some other directory where we can copy the app on Linux (at least in TaskCluster, but ideally we'd have a general solution). Alternately, we could check appPackageResult.pkg.bin to ensure it's a reasonable value before trying to use it. Or we could remove the read-pkg-up functionality, although presumably it's useful, so that's a last resort.

[Further reading](https://stackoverflow.com/questions/38747082/docker-container-has-no-tmpdir-environmental-variable) suggests that this is probably because the Docker image has no TMPDIR environment variable, which we can fix with an ENV step in the Dockerfile. However, currently we're using the stock node:6 Docker image, and I'd like to avoid building a custom image if possible.

I've tried a couple of other ways of setting TMPDIR over in #84, but so far no luck.

**Comment by mykmelez** [comment]
@cvan Figured out and fixed in https://github.com/cvan/qbrt/pull/2.

**Comment by cvan** [comment]
Thanks for the explanation. I completely missed that; that totally makes sense. Looking at PR.

File: test/package-main-path.js [hunk 1]
```diff
@@ -55,8 +55,9 @@ new Promise((resolve, reject) => {
 
   child.stderr.on('data', data => {
     const error = data.toString('utf8');
+    // console.error('test failure due to stderr output');
     console.error(error);
-    reject(error);
+    // reject(error);
```
Comment (mykmelez) [inline test/package-main-path.js]: Now that you've removed the "reading package" messages and "missing package.json" warnings, it should be possible to re-enable this error message and test failure on output to stderr.

Comment (mykmelez) [inline test/package-main-path.js]: (You should now be able to do this in the two other places where I disabled that error message and test failure.)

File: test/run-path-bad-package-name.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  });
+
+  child.stderr.on('data', data => {
+    const error = data.toString('utf8');
+    console.error(error);
+    totalError += error.trim();
+    // if (outputRegex.test(totalError) && !quitting) {
+    //   child.kill('SIGINT');
+    //   quitting = true;
+    // }
```
Comment (mykmelez) [inline test/run-path-bad-package-name.js]: I guess this commented-out code could just be removed, as it isn't needed.

File: test/package-main-path.js (hunk 1 above)
Comment (mykmelez) [inline test/package-main-path.js]: I've made these changes in 556b8a9b749e04439ec6f3224d2f982b1f60a94a.

File: test/run-path-bad-package-name.js (hunk 2 above)
Comment (mykmelez) [inline test/run-path-bad-package-name.js]: I've made this change in 834ec3c1415916b9875d1c1f19c92752000bbeea. I also made a minor change to this function in a82416ed0e9ba54997721e8dfa853e469df3132e.

**Comment by mykmelez** [comment]
I've fixed up the last few nits, so I'll merge this with those changes.

**Comment by cvan** [comment]
thanks for making these changes. sorry, I've been away working on a different project for a bit. your help's much appreciated 👍

**Pull request merged**